### PR TITLE
Added cpp ReverseModel

### PR DESCRIPTION
**Pull request by FloVanGH**

### CODE REVIEW

**Review by ogoffart — APPROVED** [review]

File: api/cpp/include/slint.h [hunk 1]
```diff
@@ -1314,6 +1315,89 @@ class SortModel : public Model<ModelData>
...
+
+    void set_row_data(size_t i, const ModelData &value) override
+    {
+        auto count = inner->source_data->row_count();
+        inner->source_model->set_row_data(count - i - 1, value);
+    }
+
+    /// Re-applies the model's sort function on each row of the source model. Use this if state
+    /// external to the sort function has changed.
+    void reset() { inner->reset(); }
```
Comment (tronical) [inline api/cpp/include/slint.h:1392]: This calls `inner->reset()`, which calls `target_model.reset()`, where `target_model` is a `ReverseModel<ModelData>`, which is the class that defines this method. Wouldn't this be an endless recursion?

IMO this function is not needed.

File: api/cpp/include/slint.h [hunk 2]
```diff
@@ -1314,6 +1315,89 @@ class SortModel : public Model<ModelData>
...
+
+    /// Returns the source model of this filter model.
```
Comment (tronical) [inline api/cpp/include/slint.h:1394]:
```suggestion
    /// Returns the source model of this reserve model.
```

File: api/cpp/include/slint.h [hunk 3]
```diff
@@ -1314,6 +1315,89 @@ class SortModel : public Model<ModelData>
...
+    }
+
+    void reset() override { target_model.reset(); }
+
+    std::shared_ptr<slint::Model<ModelData>> source_model;
+    slint::ReverseModel<ModelData> &target_model;
+};
+}
+
+/// The ReverseModel acts as an adapter model for a given source model by reverse all rows.
```
Comment (tronical) [inline api/cpp/include/slint.h:1361]:
```suggestion
/// The ReverseModel acts as an adapter model for a given source model by reserving all rows.
/// This means that the first row in the source model is the last row of this model, the second
/// row is the second last, and so on.
```

File: api/cpp/tests/models.cpp [hunk 4]
```diff
@@ -399,3 +399,101 @@ SCENARIO("Sorted Model Change")
     REQUIRE(sorted_model->row_data(2) == 2);
     REQUIRE(sorted_model->row_data(3) == 3);
 }
+
+SCENARIO("Reverse Model Insert")
+{
+    auto vec_model = std::make_shared<slint::VectorModel<int>>(std::vector<int> { 3, 4, 1, 2 });
+
+    auto reverse_model = std::make_shared<slint::ReverseModel<int>>(
+            vec_model, [](auto lhs, auto rhs) { return lhs < rhs; });
```
Comment (ogoffart) [inline api/cpp/tests/models.cpp:408]: You shouldn't pass a lambda here.  Copy-pasted from sort model?

**Pull request merged**